src/converter/jfr2pprof.java

```diff
@@ -0,0 +1,243 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import java.io.File;
+import java.io.FileOutputStream;
+import java.nio.charset.StandardCharsets;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import one.jfr.ClassRef;
+import one.jfr.Dictionary;
+import one.jfr.JfrReader;
+import one.jfr.MethodRef;
+import one.jfr.StackTrace;
+import one.jfr.event.ExecutionSample;
+
+import one.proto.Proto;
+
+/**
+ * Convert a JFR file to pprof
+ * <p>
+ * Protobuf definition: https://github.com/google/pprof/blob/44fc4e887b6b0cfb196973bcdb1fab95f0b3a75b/proto/profile.proto
+ */
+public class jfr2pprof {
+
+    public class Method {
+        public final byte[] name;
+
+        public Method(final byte[] name) {
+            this.name = name;
+        }
+
+        @Override
+        public int hashCode() {
+            return Arrays.hashCode(name);
+        }
+
+        @Override
+        public boolean equals(final Object other) {
+            if (null == other) {
+                return false;
+            }
+
+            if (other instanceof Method) {
+                final Method that = (Method) other;
+                return Arrays.equals(this.name, that.name);
+            } else {
+                return false;
+            }
+        }
+    }
+
+    private final JfrReader reader;
+
+    public jfr2pprof(final JfrReader reader) {
+        this.reader = reader;
+    }
+
+    public static final byte[] METHOD_UNKNOWN = "[unknown]".getBytes();
+
+    // Profile IDs
+    public static final int PROFILE_SAMPLE_TYPE = 1;
+    public static final int PROFILE_SAMPLE = 2;
+    public static final int PROFILE_LOCATION = 4;
+    public static final int PROFILE_FUNCTION = 5;
+    public static final int PROFILE_STRING_TABLE = 6;
+    public static final int PROFILE_TIME_NANOS = 9;
+    public static final int PROFILE_DURATION_NANOS = 10;
+    public static final int PROFILE_COMMENT = 13;
+    public static final int PROFILE_DEFAULT_SAMPLE_TYPE = 14;
+
+    // ValueType IDs
+    public static final int VALUETYPE_TYPE = 1;
+    public static final int VALUETYPE_UNIT = 2;
+
+    // Sample IDs
+    public static final int SAMPLE_LOCATION_ID = 1;
+    public static final int SAMPLE_VALUE = 2;
+
+    // Location IDs
+    public static final int LOCATION_ID = 1;
+    public static final int LOCATION_LINE = 4;
+
+    // Line IDs
+    public static final int LINE_FUNCTION_ID = 1;
+
+    // Function IDs
+    public static final int FUNCTION_ID = 1;
+    public static final int FUNCTION_NAME = 2;
+
+    // `Proto` instances are mutable, careful with reordering
+    public void dump(final FileOutputStream out) throws Exception {
```

Comment:
I'd change the type of `out` to `OutputStream`, as this method is not specific to `FileOutputStream`.